src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -1642,79 +1656,59 @@ private bool TryProcessCompilationCompleted(CompilationCompletedEvent endEvent,
         {
             return (diagnostic, analyzer) =>
             {
-                var filteredDiagnostic = GetFilteredDiagnostic(diagnostic, compilation, analyzerOptions);
+                var filteredDiagnostic = GetFilteredDiagnostic(diagnostic, compilation, analyzerOptions, severityFilter);
                 if (filteredDiagnostic != null)
                 {
                     addDiagnosticCore(filteredDiagnostic, analyzer);
                 }
             };
         }
 
-        private static Diagnostic GetFilteredDiagnostic(Diagnostic diagnostic, Compilation compilation, AnalyzerOptions analyzerOptions)
+        private static Diagnostic GetFilteredDiagnostic(Diagnostic diagnostic, Compilation compilation, AnalyzerOptions analyzerOptions, SeverityFilter severityFilter)
         {
             diagnostic = compilation.Options.FilterDiagnostic(diagnostic);
 
             // Apply bulk configuration from analyzer options for analyzer diagnostics, if applicable.
             var tree = diagnostic?.Location.SourceTree;
-            if (tree == null || analyzerOptions == null || diagnostic.CustomTags.Contains(WellKnownDiagnosticTags.Compiler) || diagnostic.IsNotConfigurable())
+            if (tree != null &&
+                analyzerOptions.TryGetSeverityFromBulkConfiguration(tree, compilation, diagnostic.Descriptor, out ReportDiagnostic severity))
             {
-                return diagnostic;
+                diagnostic = diagnostic.WithReportDiagnostic(severity);
             }
 
-            // If user has explicitly configured severity for this diagnostic ID, that should be respected and
-            // bulk configuration should not be applied.
-            // For example, 'dotnet_diagnostic.CA1000.severity = error'
-            if (compilation.Options.SpecificDiagnosticOptions.ContainsKey(diagnostic.Id) ||
-                tree.DiagnosticOptions.ContainsKey(diagnostic.Id))
-            {
-                return diagnostic;
-            }
-
-            var analyzerConfigOptions = analyzerOptions.AnalyzerConfigOptionsProvider.GetOptions(tree);
-
-            // Otherwise, if user has explicitly configured default severity for the diagnostic category, that should be respected.
-            // For example, 'dotnet_analyzer_diagnostic.category-security.severity = error'
-            var categoryBasedKey = AnalyzerConfigOptionNames.GetCategoryBasedDotnetAnalyzerDiagnosticSeverityKey(diagnostic.Category);
-            if (analyzerConfigOptions.TryGetValue(categoryBasedKey, out var value) &&
-                AnalyzerConfigSet.TryParseSeverity(value, out ReportDiagnostic severity))
+            if (!severityFilter.IsEmpty &&
+                diagnostic != null &&
+                severityFilter.Contains(DiagnosticDescriptor.MapSeverityToReport(diagnostic.Severity)))
```

Comment:
What if `diagnostic.Severity` is `Suppress` here? The other place where we run `severityFilter.Contains`  we do an explicit check for `Suppress` as well.

Reply:
`Diagnostic.Severity` is actually `DiagnosticSeverity` enum which does not have a `Suppress` field: https://github.com/dotnet/roslyn/blob/d4bf6990e488e58dd97b835284eaf731731df799/src/Compilers/Core/Portable/Diagnostic/DiagnosticSeverity.cs#L9-L35

Other places we are using `ReportDiagnostic`, which has `Suppress` field: https://github.com/dotnet/roslyn/blob/d4bf6990e488e58dd97b835284eaf731731df799/src/Compilers/Core/Portable/Diagnostic/ReportDiagnostic.cs#L9-L43